smbios/internal/github.com/digitalocean/go-smbios/smbios/decoder.go

```diff
@@ -0,0 +1,253 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smbios
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+const (
+	// headerLen is the length of the Header structure.
+	headerLen = 4
+
+	// typeEndOfTable indicates the end of a stream of Structures.
+	typeEndOfTable = 127
+)
+
+var (
+	// Byte slices used to help parsing string-sets.
+	null         = []byte{0x00}
+	endStringSet = []byte{0x00, 0x00}
+)
+
+// A Decoder decodes Structures from a stream.
+type Decoder struct {
+	br *bufio.Reader
+	b  []byte
+}
+
+// Stream locates and opens a stream of SMBIOS data and the SMBIOS entry
+// point from an operating system-specific location.  The stream must be
+// closed after decoding to free its resources.
+//
+// If no suitable location is found, an error is returned.
+func Stream() (io.ReadCloser, EntryPoint, error) {
+	rc, ep, err := stream()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// The io.ReadCloser from stream could be any one of a number of types
+	// depending on the source of the SMBIOS stream information.
+	//
+	// To prevent the caller from potentially tampering with something dangerous
+	// like mmap'd memory by using a type assertion, we make the io.ReadCloser
+	// into an opaque and unexported type to prevent type assertion.
+	return &opaqueReadCloser{rc: rc}, ep, nil
+}
+
+// NewDecoder creates a Decoder which decodes Structures from the input stream.
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{
+		br: bufio.NewReader(r),
+		b:  make([]byte, 1024),
+	}
+}
+
+// Decode decodes Structures from the Decoder's stream until an End-of-table
+// structure is found.
+func (d *Decoder) Decode() ([]*Structure, error) {
+	var ss []*Structure
+
+	for {
+		s, err := d.next()
+		if err != nil {
+			return nil, fmt.Errorf("next error: %w", err)
+		}
+
+		// End-of-table structure indicates end of stream.
+		ss = append(ss, s)
+
+		if s.Header.Type == typeEndOfTable {
+			break
+		}
+	}
+
+	return ss, nil
+}
+
+// next decodes the next Structure from the stream.
+func (d *Decoder) next() (*Structure, error) {
+	h, err := d.parseHeader()
+	if err != nil {
+		return nil, err
+	}
+
+	// Length of formatted section is length specified by header, minus
+	// the length of the header itself.
+	l := int(h.Length) - headerLen
+
+	// Hyper-V SMBIOS tables seem to be broken, implement a workaround
+	if h.Length == 0 && h.Type == 0 {
+		return &Structure{
+			Header: Header{
+				Type: typeEndOfTable,
+			},
+		}, nil
+	}
```

Comment:
Is this the actual fix?

Reply:
yes, sorry, I couldn't put that a single commit in a more clear way.